assets/js/components/Calculator.js

```diff
@@ -0,0 +1,177 @@
+     */
+    const handleSwapCurrencies = () => {
+        const tempCurrency = fromCurrency;
+        setFromCurrency(toCurrency);
+        setToCurrency(tempCurrency);
+    };
+
+    // Get exchange rate information between the selected currencies
+    const getExchangeRateInfo = () => {
+        const fromRate = rates.find(rate => rate.code === fromCurrency)?.nbpRate || 1;
+        const toRate = rates.find(rate => rate.code === toCurrency)?.nbpRate || 1;
+        const exchangeRate = (fromRate / toRate).toFixed(4);
+        const reverseRate = (toRate / fromRate).toFixed(4);
+
+        return (
+            <>
+                <p className="m-0 p-0">1 {fromCurrency} = {exchangeRate} {toCurrency}</p>
+                <p className="m-0 p-0">1 {toCurrency} = {reverseRate} {fromCurrency}</p>
+            </>
+        );
+    };
+
+    // Automatically convert currencies when amount, source currency, or target currency changes
+    useEffect(() => {
+        handleConversion();
+    }, [amount, fromCurrency, toCurrency]);
+
+    return (
+        <div className="container mt-4">
+            <h2>Currency Calculator</h2>
+            <p>Exchange rates as of: <strong>{dataFromTheDay}</strong></p> {/* Display the date of exchange rate */}
+
+            {/* First line: 'I HAVE' (input) + currency */}
+            <div className="row">
+                <div className="col-3">
+                    <label htmlFor="amount">I HAVE</label>
+                    <input
+                        type="number"
+                        id="amount"
+                        name="amount"
+                        className="form-control"
+                        value={amount}
+                        onChange={handleAmountChange}
+                        placeholder="Enter amount"
+                    />
+                </div>
+                <div className="col-2">
+                    <label htmlFor="fromCurrency">&nbsp;</label>
+                    <select
+                        id="fromCurrency"
+                        name="fromCurrency"
+                        className="form-control"
+                        value={fromCurrency}
+                        onChange={(e) => {
+                            setFromCurrency(e.target.value);
+                            if (e.target.value === toCurrency) {
+                                setToCurrency(rates.find(rate => rate.code !== e.target.value).code); // Automatically change target currency if they are the same
+                            }
+                        }}
+                    >
+                        {rates
+                            .filter(rate => rate.code !== toCurrency) // Hide the currency that is currently selected in 'To'
+                            .map(rate => (
+                                <option key={rate.code} value={rate.code}>
+                                    {rate.code}
+                                </option>
+                            ))}
```

Comment:
Czy możemy przenieść map/filter do nowej tablicy, żeby wywoływać ją wyżej przed renderowaniem JSX? Chodzi o to, żeby bardziej odseparować logikę js od kodu html. Można stworzyć np. array filteredRates, na którym zrobisz filter + map, a poniżej w kodzie już bezpośrednio użyć filteredRates